### PR TITLE
Move QueryState from ExecutableNetwork to InferRequest

### PR DESCRIPTION
**Pull request by sadolini**
ExecutableNetwork::QueryState should be deprecated. Also minor changes done in MemoryState API: GetLastState renamed to GetState, added implementation of this method on CPU. GetName() on CPU now returns variable_id (as in GNA already done). Description of State API in comments fixed.

### CODE REVIEW

File: inference-engine/include/cpp/ie_memory_state.hpp [hunk 1]
```diff
@@ -75,4 +80,5 @@ class MemoryState {
     }
 };
 
+using VariableState = MemoryState;
```
Comment (ilya-lavrenov) [inline inference-engine/include/cpp/ie_memory_state.hpp:83]: can we rename `MemoryState` to `VariableState` and make the following alias:
```c++
/*
 * @brief For compatibility reasons.
 */
using MemoryState = VariableState;
```

File: inference-engine/include/ie_iinfer_request.hpp [hunk 2]
```diff
@@ -177,6 +178,20 @@ class IInferRequest : public details::IRelease {
      * @return Enumeration of the resulted action: InferenceEngine::OK (0) for success
      */
     virtual InferenceEngine::StatusCode SetBatch(int batch_size, ResponseDesc* resp) noexcept = 0;
+
+    /**
+    * @brief Gets state control interface for given executable network.
```
Comment (ilya-lavrenov) [inline inference-engine/include/ie_iinfer_request.hpp:183]: `given executable network` -> `given inference request`

File: inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_async_request_internal.hpp [hunk 3]
```diff
@@ -59,6 +59,11 @@ class AsyncInferRequestInternal : public IAsyncInferRequestInternal, public Infe
         StartAsyncImpl();
     };
 
+    std::vector<IMemoryStateInternal::Ptr> QueryState() override {
+        // meaning base plugin reports as no state available - plugin owners need to create proper override of this
+        return {};
```
Comment (ilya-lavrenov) [inline inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_async_request_internal.hpp:64]: is not it better to throw an exception?

File: inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_async_request_thread_safe_internal.hpp [hunk 4]
```diff
@@ -131,6 +131,11 @@ class AsyncInferRequestThreadSafeInternal : public IAsyncInferRequestInternal {
         SetBatch_ThreadUnsafe(batch);
     };
 
+    std::vector<IMemoryStateInternal::Ptr> QueryState() override {
+        // meaning base plugin reports as no state available - plugin owners need to create proper override of this
+        return {};
```
Comment (ilya-lavrenov) [inline inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_async_request_thread_safe_internal.hpp:136]: the same

File: inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_request_internal.hpp [hunk 5]
```diff
@@ -254,6 +254,25 @@ class InferRequestInternal : virtual public IInferRequestInternal {
...
+    InferenceEngine::BlobMap _inputs;  //!< A map of network input blobs
+    InferenceEngine::BlobMap _outputs;  //!< A map of network output blobs
+    std::map<std::string, PreProcessDataPtr> _preProcData;  //!< A map of pre-process data per input
+    int m_curBatch;  //!< Current batch value used in dynamic batching
+
+    /**
+     * @brief A shared pointer to ExecutableNetworkInternal interface
+     * @note Needed to correctly handle ownership between objects.
+     */
+    std::shared_ptr<ExecutableNetworkInternal> _exeNetwork;
```
Comment (ilya-lavrenov) [inline inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_request_internal.hpp:274]: looks like all these fields are duplicated. Should we remove them? They are already a part of this interface.

**Comment by ilya-lavrenov** [comment]
We also need to update `docs/IE_DG/API_Changes.md`

File: inference-engine/include/cpp/ie_memory_state.hpp [hunk 6]
```diff
@@ -58,12 +58,17 @@ class MemoryState {
      * Wraps IMemoryState::GetLastState
```
Comment (ilya-lavrenov) [inline inference-engine/include/cpp/ie_memory_state.hpp:58]: we also need to add second optional parameter to `explicit MemoryState(IMemoryState::Ptr pState)`:
```c++
explicit MemoryState(IMemoryState::Ptr pState, details::SharedObjectLoader::Ptr plg = {})
```
And need to add a test for such case (see `inference-engine/tests/functional/plugin/shared/src/behavior/cpp_holders.cpp` as a reference)

File: inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_request_internal.hpp (hunk 5 above)
Comment (sadolini) [reply inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_request_internal.hpp]: OMG, it definitely should not be here

File: inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_async_request_internal.hpp (hunk 3 above)
Comment (sadolini) [reply inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_async_request_internal.hpp:65]: agree, will fix

File: inference-engine/include/cpp/ie_memory_state.hpp (hunk 1 above)
Comment (sadolini) [reply inference-engine/include/cpp/ie_memory_state.hpp]: ok

File: inference-engine/include/cpp/ie_memory_state.hpp (hunk 6 above)
Comment (sadolini) [reply inference-engine/include/cpp/ie_memory_state.hpp]: Ctor was updated, test extending is in progress

File: inference-engine/include/ie_iinfer_request.hpp [hunk 7]
```diff
@@ -177,6 +178,18 @@ class IInferRequest : public details::IRelease {
...
+    *
+    * State control essential for recurrent networks
+    *
+    * @param pState reference to a pointer that receives internal states
+    * @param idx requested index for receiving memory state
+    * @param resp Optional: pointer to an already allocated object to contain information in case of failure
+    * @return Status code of the operation: InferenceEngine::OK (0) for success, OUT_OF_BOUNDS (-6) no memory state for
+    * given index
+    */
+    virtual StatusCode QueryState(IMemoryState::Ptr& pState, size_t idx, ResponseDesc* resp) noexcept = 0;
```
Comment (ilya-lavrenov) [inline inference-engine/include/ie_iinfer_request.hpp:193]: `IMemoryState` -> `IVariableState`

File: inference-engine/src/gna_plugin/gna_executable_network.hpp [hunk 8]
```diff
@@ -59,12 +59,13 @@ class GNAExecutableNetwork : public InferenceEngine::ExecutableNetworkThreadSafe
         return std::make_shared<GNAInferRequest>(plg, networkInputs, networkOutputs);
     }
 
-
-
+    INFERENCE_ENGINE_DEPRECATED("Use InferRequest::QueryState instead")
     std::vector<InferenceEngine::IMemoryStateInternal::Ptr>  QueryState() override {
```
Comment (ilya-lavrenov) [inline inference-engine/src/gna_plugin/gna_executable_network.hpp:63]: can we replace `IMemoryState*` with `IVariableState*` everywhere? except `using IMemoryState = IVariableState;`

Comment (sadolini) [reply inference-engine/src/gna_plugin/gna_executable_network.hpp]: replaced *MemoryState* to *VariableState* everywhere (except constructions like IMemoryState=IVariableState of course)

File: inference-engine/include/ie_iinfer_request.hpp (hunk 7 above)
Comment (sadolini) [reply inference-engine/include/ie_iinfer_request.hpp]: done

**Comment by sadolini** [comment]
@AlexPeskov  please review cpu part
@dorloff  please review gna part

**Review by ilya-lavrenov — APPROVED** [review]

**Review by AlexPeskov — APPROVED** [review]
LGTM

**Review by ilyachur — APPROVED** [review]

**Review by dorloff — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
@sadolini thanks!

**Pull request merged**